### PR TITLE
Minimize error in space usage accounting

### LINKED ISSUE #463: Reactive disk usage accounting can lead to breaching cache size restriction for brief periods
Today the way disk space utilization by a file, and hence the full Cache, is calculated is when the client calls the `setAllCached` api to inform BookKeeper that new data was cached by it. In this api handler, we update the disk usage of the file.

Consider a scenario where we already have 100% utilization of the space configured for Rubix. Now a new file is read, say 100MB in one inputStream by client. RemoteRRC or FileDownloadRRC would cache that 100MB first and then inform BookKeeper via `setAllCached` about the new data. BookKeeper will update the cache entry of this file with new weight=100MB and then realize that evictions are needed to accommodate this data. It will then do necessary evictions but for some time the disk usage by Rubix will go over configured percentage.

We can fix this by updating weight of the file in `getCacheStatus` api itself. When this api is returning `LOCAL` blockStatus then we know that client would eventually bring those blocks in. We could use this info to pro-actively update the weight of the file.

Now there might be a mis-behaving client that calls getCacheStatus but never downloads the data, e.g. a task cancellation comes before the download could start. In that case we would be over estimating the disk usage. One way to fix this is to have a scheduled thread to update file weights periodically based on usage on disk, this would ensure that realtime tracking is conservative and will never cross the configured threshold and we fixed the under-estimations by the scheduled service.

### PR DESCRIPTION
**Pull request by shubhamtagra**
Fixes #463 

Main changes:
1. Limit the size of each readRequest in FileDownloadRequestChain to 100MB
2. With the download of each readRequest itself update the metadata instead of existing behavior of downloading all the data in the Chain and then updating metadata

This ensures we will never cross over the configured threshold of disk utilization by 1GB. This crossing over is actually partially handled by only considering 95% of disk space as available, keeping 5% as buffer space for this cross over but in existing warmup scenario we could cross the limit by a much bigger margin.

### CODE REVIEW

File: rubix-core/src/main/java/com/qubole/rubix/core/ReadRequestChain.java [hunk 1]
```diff
@@ -46,19 +48,32 @@
...
+  }
+
   // Should be added in forward seek fashion for better performance
   public void addReadRequest(ReadRequest readRequest)
   {
     checkState(!isLocked, "Adding request to a locked chain");
     log.debug(String.format("Request to add ReadRequest: [%d, %d, %d, %d, %d]", readRequest.getBackendReadStart(), readRequest.getBackendReadEnd(), readRequest.getActualReadStart(), readRequest.getActualReadEnd(), readRequest.getDestBufferOffset()));
     if (lastRequest == null) {
-      addRequest(readRequest);
+      lastRequest = readRequest;
```
Comment (harmandeeps) [inline rubix-core/src/main/java/com/qubole/rubix/core/ReadRequestChain.java:75]: can you please explain this? with this change, we are adding the last request in the `lock()` method.

Comment (stagraqubole) [reply rubix-core/src/main/java/com/qubole/rubix/core/ReadRequestChain.java:75]: Yes, earlier we used to keep the reference to lastRequest put on queue and collated new request with it if they were sequential. That was not clean if were to add the logic of splitting the request by size. Doing it this way was cleaner as the most recent readRequest goes into kind of a staging space and can be updated but ones gone into queue are immutable.

File: rubix-spi/src/main/java/com/qubole/rubix/spi/CacheConfig.java [hunk 2]
```diff
@@ -108,6 +108,7 @@
   private static final String KEY_DUMMY_MODE = "rubix.cache.dummy.mode";
   private static final String KEY_EMBEDDED_MODE = "rubix.cluster.embedded.mode";
   private static final String KEY_HEARTBEAT_ENABLED = "rubix.cluster.heartbeat.enabled";
+  private static final String KEY_PARALLEL_WARMUP_MAX_CHUNK_SIZE = "rubix.cache.parallel.warmup.max_chunk_size";
```
[the remark this answers is not in the document]
Comment (stagraqubole) [reply rubix-spi/src/main/java/com/qubole/rubix/spi/CacheConfig.java:111]: This wont have `min` counter part, so we dont need the hierarchy.

File: rubix-core/src/main/java/com/qubole/rubix/core/ReadRequestChain.java [hunk 3]
```diff
@@ -69,22 +84,47 @@ public void addReadRequest(ReadRequest readRequest)
...
 
   private void addRequest(ReadRequest readRequest)
+  {
+    if (readRequest.getBackendReadLength() <= maxReadRequestSize) {
+      addRequestToQueue(readRequest);
+      return;
+    }
+
+    long start = 0;
+    while (start < readRequest.getBackendReadLength()) {
```
[the remark this answers is not in the document]
Comment (stagraqubole) [reply rubix-core/src/main/java/com/qubole/rubix/core/ReadRequestChain.java]: done

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/FileDownloadRequestChain.java [hunk 4]
```diff
@@ -201,22 +212,19 @@ private void writeToFile(byte[] buffer, int length, FileChannel fileChannel, lon
...
+      bookKeeper.setAllCached(remotePath,
+              fileSize,
+              lastModified,
+              toStartBlock(readRequest.getBackendReadStart(), blockSize),
+              toEndBlock(readRequest.getBackendReadEnd(), blockSize),
+              generationNumber);
     }
     catch (Exception e) {
-      log.debug("Could not update BookKeeper about newly cached blocks: " + Throwables.getStackTraceAsString(e));
+      log.warn(String.format("Unable to update cache status for %s:%d:%d, this can cause wrong accounting of disk utilization"), e);
```
[the remark this answers is not in the document]
Comment (stagraqubole) [reply rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/FileDownloadRequestChain.java]: fixed

**Pull request merged**